## Homework #1

#### Problem 4.2

In [11]:
# Problem 4.2: Generate 3 random 3x3 matrices and compute inverses and product

import numpy as np

np.random.seed()  

matrices = []
inverses = []
products = []

for i in range(3):
    while True:
        M = np.random.randint(-10, 11, size=(3,3))
        det = np.linalg.det(M)
        if abs(det) > 1e-8:  
            break
    matrices.append(M)
    inv_M = np.linalg.inv(M)
    inverses.append(inv_M)
    products.append(np.dot(M, inv_M))

# Display results
for i in range(3):
    print(f"\nMatrix {i+1}:")
    print(matrices[i])
    
    print("\nInverse:")
    print(inverses[i])
    
    print("\nProduct (Matrix * Inverse):")
    print(products[i])
    print("\n" + "-"*50)



Matrix 1:
[[-3 10  3]
 [-3 -8 -7]
 [ 5 -8  8]]

Inverse:
[[-0.27149321 -0.23529412 -0.1040724 ]
 [-0.02488688 -0.08823529 -0.0678733 ]
 [ 0.14479638  0.05882353  0.12217195]]

Product (Matrix * Inverse):
[[ 1.00000000e+00 -8.32667268e-17  6.93889390e-17]
 [ 5.55111512e-17  1.00000000e+00 -1.38777878e-17]
 [ 0.00000000e+00  0.00000000e+00  1.00000000e+00]]

--------------------------------------------------

Matrix 2:
[[ 6 -5 10]
 [ 7  8  8]
 [ 1  3 -3]]

Inverse:
[[ 0.15841584 -0.04950495  0.3960396 ]
 [-0.09570957  0.09240924 -0.07260726]
 [-0.04290429  0.07590759 -0.27392739]]

Product (Matrix * Inverse):
[[ 1.00000000e+00  8.32667268e-17  1.11022302e-16]
 [-5.55111512e-17  1.00000000e+00  4.44089210e-16]
 [ 6.93889390e-18 -9.71445147e-17  1.00000000e+00]]

--------------------------------------------------

Matrix 3:
[[-6  5 10]
 [-5  8 -7]
 [ 0 10 -5]]

Inverse:
[[-0.03726708 -0.1552795   0.14285714]
 [ 0.0310559  -0.03726708  0.11428571]
 [ 0.0621118  -0.07453416  0.02857143]]

P

#### Problem 5.1

In [16]:
import pandas as pd

# Load dataset 
df = pd.read_csv("kc_house_data.csv")

# Drop columns to ignore
df = df.drop(columns=['id', 'date', 'zipcode'])

# Compute statistics
stats = pd.DataFrame({
    'mean': df.mean(),
    'min': df.min(),
    'max': df.max(),
    'variance': df.var()
})

print("Feature statistics (average, min, max, variance):")
print(stats)

# Find feature with lowest and highest average
lowest_avg_feature = stats['mean'].idxmin()
highest_avg_feature = stats['mean'].idxmax()

print("\nLowest average:")
print(f"{lowest_avg_feature}: {stats.loc[lowest_avg_feature, 'mean']}")

print("\nHighest average:")
print(f"{highest_avg_feature}: {stats.loc[highest_avg_feature, 'mean']}")

# Find feature with lowest and highest variance
lowest_var_feature = stats['variance'].idxmin()
highest_var_feature = stats['variance'].idxmax()

print("\nLowest variance:")
print(f"{lowest_var_feature}: {stats.loc[lowest_var_feature, 'variance']}")

print("\nHighest variance:")
print(f"{highest_var_feature}: {stats.loc[highest_var_feature, 'variance']}")


Feature statistics (average, min, max, variance):
                        mean         min           max      variance
price          540088.141767  75000.0000  7.700000e+06  1.347824e+11
bedrooms            3.370842      0.0000  3.300000e+01  8.650150e-01
bathrooms           2.114757      0.0000  8.000000e+00  5.931513e-01
sqft_living      2079.899736    290.0000  1.354000e+04  8.435337e+05
sqft_lot        15106.967566    520.0000  1.651359e+06  1.715659e+09
floors              1.494309      1.0000  3.500000e+00  2.915880e-01
waterfront          0.007542      0.0000  1.000000e+00  7.485226e-03
view                0.234303      0.0000  4.000000e+00  5.872426e-01
condition           3.409430      1.0000  5.000000e+00  4.234665e-01
grade               7.656873      1.0000  1.300000e+01  1.381703e+00
sqft_above       1788.390691    290.0000  9.410000e+03  6.857347e+05
sqft_basement     291.509045      0.0000  4.820000e+03  1.958727e+05
yr_built         1971.005136   1900.0000  2.015000e+0

#### Problem 5.2

In [22]:
# Compute correlation of each feature with the response 

import pandas as pd

# Load dataset
df = pd.read_csv('kc_house_data.csv')

# Drop columns to ignore
df = df.drop(columns=['id', 'date', 'zipcode'])

# Compute correlations with price
correlations = df.corr()['price'].drop('price')

# Convert to DataFrame for clean table
corr_table = correlations.sort_values(ascending=False).reset_index()
corr_table.columns = ['Feature', 'Correlation_with_Price']

corr_table


Feature  Correlation_with_Price
0     sqft_living                0.702035
1           grade                0.667434
2      sqft_above                0.605567
3   sqft_living15                0.585379
4       bathrooms                0.525138
5            view                0.397293
6   sqft_basement                0.323816
7        bedrooms                0.308350
8             lat                0.307003
9      waterfront                0.266369
10         floors                0.256794
11   yr_renovated                0.126434
12       sqft_lot                0.089661
13     sqft_lot15                0.082447
14       yr_built                0.054012
15      condition                0.036362
16           long                0.021626

All of the features have positive correlation coefficients, which means as the feature increases, on average house price tends to increase. The feature which has the highest positive correlation is sqft_living at 0.7020. This means this feature is the most important predictor in terms of linear relationship with price.

#### Problem 5.3

All computed correlation coefficients were positive, meaning that as each feature increases, the house price tends to increase even if the relationship is more weak such as long, condition, and yr_built.